records_mover/url/optimizer/__init__.py

```diff
@@ -0,0 +1,121 @@
+from contextlib import contextmanager
+from records_mover.url.base import BaseDirectoryUrl
+import logging
+from .gcp_data_transfer_service import GCPDataTransferService
+from typing import Iterator, Tuple, TYPE_CHECKING
+if TYPE_CHECKING:
+    from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+    from records_mover.url.filesystem import FilesystemDirectoryUrl
+
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: Break out GCP stuff and AWS stuff into separate classes
+class CopyOptimizer:
+    """When Records Mover needs to move data between different URL
+    schemes, the default way (a single-threaded downloading of bytes
+    from the source and then reuploading of bytes to the target) can
+    be very slow for large directories.
+
+    This class can use faster means (ideally a cloud-based service
+    that can copy things at scale, or at least vendor-optimized
+    multi-threaded upload/download code) when they are available.
+
+    This depends on the specifics of the copy, and as such return
+    values must be checked to determine if the copy was able to be
+    performed.
+    """
+
+    def __init__(self) -> None:
+        # TODO: can we use method names to start odwn the path of a stronger abstraction?
+        self._gcp_data_transfer = GCPDataTransferService()
+
+    def copy(self,
+             loc: BaseDirectoryUrl,
+             other_loc: BaseDirectoryUrl) -> bool:
+        """Copy from one directory to another using optimized means.
+
+        :return: True if the copy was performed, or False if no
+        optimized means were possible."""
+        if loc.scheme == 's3' and other_loc.scheme == 'file':
+            from records_mover.url.filesystem import FilesystemDirectoryUrl
+            from records_mover.url.s3.s3_directory_url import S3DirectoryUrl
+
+            assert isinstance(loc, S3DirectoryUrl)
+            assert isinstance(other_loc, FilesystemDirectoryUrl)
+            # TODO: Also factor out this class
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker